File: rb/D1.rb

def parseInput()
    numbers = []
    File.open("../inputs/D1.input", "r") do |f|
        f.each_line do |line|
          numbers.append(Integer(line))
        end
      end
    return numbers
end

def p1()
    numbers = parseInput().sort()
    left = 0
    right = numbers.length() - 1
    while numbers[left] + numbers[right] != 2020
      if numbers[left] + numbers[right] > 2020
        right -= 1
      else
        left += 1
      end
    end
    return numbers[left] * numbers[right]
end

def p2()
  numbers = parseInput().sort()
  
  for i in 0...numbers.length()
    remainder = 2020 - numbers[i]
    left = i + 1
    right = numbers.length() - 1
    while left < right 
      if numbers[left] + numbers[right] == remainder
        return numbers[left] * numbers[right] * numbers[i]
      elsif numbers[left] + numbers[right] > remainder
        right -= 1
      else
        left += 1
      end
    end
  end
end

puts p1()
puts p2()

File: java/D1.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;

class D1 {
    public static void main(String[] args) throws Exception {
        System.out.println(p1());
        System.out.println(p2());
    }

    public static int p1() throws Exception {
        var nums = parseInput();
        Collections.sort(nums);
        var L = 0;
        var R = nums.size() - 1;
        while (nums.get(L) + nums.get(R) != 2020) {
            if (nums.get(L) + nums.get(R) < 2020)
                L++;
            else
                R--;
        }
        return nums.get(L) * nums.get(R);
    }

    public static int p2() throws Exception {
        var nums = parseInput();
        Collections.sort(nums);
        for (var i = 0; i < nums.size(); i++) {
            var remainder = 2020 - nums.get(i);
            var L = i + 1;
            var R = nums.size() - 1;
            while (L < R) {
                var left = nums.get(L);
                var right = nums.get(R);
                if (left + right == remainder)
                    return left * right * nums.get(i);
                else if (left + right < remainder)
                    L++;
                else
                    R--;
            }
        }
        return -1;
    }

    public static ArrayList<Integer> parseInput() throws Exception {
        ArrayList<Integer> numbers = new ArrayList<Integer>();
        for (var line : Files.readAllLines(Path.of("../inputs/D1.input")))
        {
            numbers.add(Integer.parseInt(line));
        }
        return numbers;
    }
}